feat: Make prometheus server address optional/centrally configured

RaviHari opened the pull request:
Signed-off-by: Ravi Hari <ravireliable@gmail.com>

Fixes: #1846 

```
- add a new cli option to pass prometheusAddress that can be centrally configured.
- if analysis template has the prometheus address that is preferred.
- it wouldn't accept conflicting addresses as the arguments are combined when analysis run is created with templates.
- if analysis template does not have the prometheus address defined the address from the cli argument is taken for analysisrun.
```

Checklist:

* [X] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-rollouts/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [X] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [X] I've signed my commits with [DCO](https://github.com/argoproj/argoproj)
* [X] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [X] My builds are green. Try syncing with master if they are not. 
* [X] My organization is added to [USERS.md](https://github.com/argoproj/argo-rollouts/blob/master/USERS.md).

harikrongali commented on .github/workflows/docker-publish.yml:
```diff
@@ -23,13 +23,13 @@ jobs:
         uses: docker/setup-buildx-action@v1
 
-      - name: Cache Docker layers
-        uses: actions/cache@v2
```
can you rebase to master? these changes are not related to your PR

harikrongali replied:
can you rebase to master? these changes are not related to your PR

harikrongali commented on Dockerfile:
```diff
@@ -3,7 +3,7 @@
 # Also used as the image in CI jobs so needs all dependencies
 ####################################################################################################
-FROM golang:1.17.6 as builder
+FROM --platform=$BUILDPLATFORM golang:1.17.6 as builder
```
same here

harikrongali replied:
same here

harikrongali commented on cmd/rollouts-controller/main.go:
```diff
@@ -245,6 +247,7 @@ func newCommand() *cobra.Command {
 	command.Flags().DurationVar(&electOpts.LeaderElectionLeaseDuration, "leader-election-lease-duration", controller.DefaultLeaderElectionLeaseDuration, "The duration that non-leader candidates will wait after observing a leadership renewal until attempting to acquire leadership of a led but unrenewed leader slot. This is effectively the maximum duration that a leader can be stopped before it is rep ...
 	command.Flags().DurationVar(&electOpts.LeaderElectionRenewDeadline, "leader-election-renew-deadline", controller.DefaultLeaderElectionRenewDeadline, "The interval between attempts by the acting master to renew a leadership slot before it stops leading. This must be less than or equal to the lease duration. This is only applicable if leader election is enabled.")
 	command.Flags().DurationVar(&electOpts.LeaderElectionRetryPeriod, "leader-election-retry-period", controller.DefaultLeaderElectionRetryPeriod, "The duration the clients should wait between attempting acquisition and renewal of a leadership. This is only applicable if leader election is enabled.")
+	command.Flags().StringVar(&prometheusAddress, "prometheus-address", "", "Set the  prometheus address that analysis controller uses when querying for metrics.")
```
what is the default address? do we set anywhere?

harikrongali replied:
what is the default address? do we set anywhere?

harikrongali commented on pkg/kubectl-argo-rollouts/info/testdata/canary/canary-rollout5.yaml:
```diff
@@ -0,0 +1,96 @@
+apiVersion: argoproj.io/v1alpha1
+kind: Rollout
+metadata:
```
is this related to your change?

harikrongali replied:
is this related to your change?

harikrongali commented on utils/defaults/defaults.go:
```diff
@@ -49,6 +49,7 @@ const (
 	DefaultSMITrafficSplitVersion       = "v1alpha1"
 	DefaultTargetGroupBindingAPIVersion = "elbv2.k8s.aws/v1beta1"
 	DefaultAppMeshCRDVersion            = "v1beta2"
+	DefaultPrometheusAddress            = "http://127.0.0.1:9090"
```
i see default here.

harikrongali replied:
i see default here.

harikrongali commented:
can you rebase to latest? i see code diff not related to prometheus change.

also, can you confirm if it is backward compatible? 

if the template has a Prometheus address, do you accept the URL?

harikrongali commented:
can you rebase to latest? i see code diff not related to prometheus change.

also, can you confirm if it is backward compatible? 

if the template has a Prometheus address, do you accept the URL?

harikrongali commented:
there seems to be rebase issue, i see commits listed in the diff rather master rebase.

harikrongali commented:
there seems to be rebase issue, i see commits listed in the diff rather master rebase.

RaviHari commented:
> there seems to be rebase issue, i see commits listed in the diff rather master rebase.

Fixed and rebased with master.. only changes from my pr are made available now.

RaviHari commented:
> can you rebase to latest? i see code diff not related to prometheus change.
> 
> also, can you confirm if it is backward compatible?
> 
> if the template has a Prometheus address, do you accept the URL?

The cli is optional if we do not pass anything it will use from the templates so yes it is backward compatible.
If the template has prometheus address it will accept that.. if it is not passed it will pickit up from the cli address provided from cli.

harikrongali commented:
got it. Thanks for the explanation. LGTM

harikrongali commented:
@agrawroh can you review as well?

jessesuen commented on cmd/rollouts-controller/main.go:
```diff
@@ -69,6 +69,7 @@ func newCommand() *cobra.Command {
 		awsVerifyTargetGroup bool
 		namespaced           bool
 		printVersion         bool
+		prometheusAddress    string
```
We should configure this via configmap/secret instead of CLI flag (just like we configure datadog, new relic, wavefront). Could we switch to that model instead?

agrawroh replied:
Hmmm. What do you think about configuring this using an environment variable instead? ConfigMap or Secret doesn't look like a good place to configure this. Maybe I'm missing some background/context.

RaviHari replied:
Looking a some k8s references: https://kubernetes.io/docs/reference/command-line-tools-reference I thought cli paramter option was good. I can update with configmap / secret.. however would like to understand why they are preferred?

agrawroh commented on cmd/rollouts-controller/main.go:
```diff
@@ -94,6 +95,7 @@ func newCommand() *cobra.Command {
 			defaults.SetAmbassadorAPIVersion(ambassadorVersion)
 			defaults.SetSMIAPIVersion(trafficSplitVersion)
 			defaults.SetAppMeshCRDVersion(appmeshCRDVersion)
+			defaults.SetPrometheusAddress(prometheusAddress)
```
Is it possible to do some validation inside `SetPrometheusAddress()` to make sure that the passed value is a URI in the expected format (Scheme, Port Number, ...)?

RaviHari commented:
@jessesuen and @agrawroh can you please review this pr updated as per the suggestions to pick the address from secret or env if not specified in template. Also validating the url format is correct.

agrawroh commented:
@jessesuen Thoughts on whether we also need to pick this up from K8S Secrets? It doesn't look right to me but if you think it's good to have it for consistency then it's okay. I do think that ENV is probably the right way to define these settings.

jessesuen commented:
> @jessesuen Thoughts on whether we also need to pick this up from K8S Secrets? It doesn't look right to me but if you think it's good to have it for consistency then it's okay. I do think that ENV is probably the right way to define these settings.

I discussed with @harikrongali. Looking back, I think we originally made a mistake on how datadog, wavefront, etc.. were configured via a Kubernetes Secret when they could have been done with environment variables (which happen to be backed by a secret).

Going forward (starting with this integration), I agree we should move to a model where metric provider connection information can be configured via environment variables, which will eliminate the need for the rollout controller to look up the kubernetes secret.

@RaviHari -- could we change the implementation once again to remove the option to look up the kubernetes secret and rely solely on env vars? The user could choose to store this as a secret or not (e.g. when we later add support for prometheus authentication).

agrawroh commented:
> > @jessesuen Thoughts on whether we also need to pick this up from K8S Secrets? It doesn't look right to me but if you think it's good to have it for consistency then it's okay. I do think that ENV is probably the right way to define these settings.
> 
> I discussed with @harikrongali. Looking back, I think we originally made a mistake on how datadog, wavefront, etc.. were configured via a Kubernetes Secret when they could have been done with environment variables (which happen to be backed by a secret).
> 
> Going forward (starting with this integration), I agree we should move to a model where metric provider connection information can be configured via environment variables, which will eliminate the need for the rollout controller to look up the kubernetes secret. Later we can improve the other implementations such that they could also be configured via env vars.
> 
> I just filed #1979 as a future improvement
> 
> @RaviHari -- could we change the implementation once again to remove the option to look up the kubernetes secret and rely solely on env vars? The user could choose to store this as a secret or not. Currently, since this is just a prometheus address, it there's no need for it to be a secret. Later, if/when we later add support for prometheus authentication, they would want it to be a secret, obviously.

@jessesuen Thanks for chiming in. I can probably take the improvement you filed if I get some free time next week.

RaviHari commented:
@jessesuen and @agrawroh thanks.. I made the changes to make read it from env variable. Can you please review.

agrawroh commented on metricproviders/prometheus/prometheus.go:
```diff
@@ -22,6 +26,8 @@ const (
 	// ResolvedPrometheusQuery is used as the key for storing the resolved prometheus query in the metrics result
 	// metadata object.
 	ResolvedPrometheusQuery = "ResolvedPrometheusQuery"
+	TokensSecretName        = "prometheus"
```
Do we still need this now?

agrawroh commented on metricproviders/prometheus/prometheus.go:
```diff
@@ -140,12 +146,51 @@ func NewPrometheusProvider(api v1.API, logCtx log.Entry) *Provider {
+	if err != nil {
+		log.Errorf("Error in parsing url: %v", err)
+	}
+	log.Infof("Parsed url: %v", u)
```
`info` might be too verbose. Should we change it to `debug` instead?

agrawroh commented on metricproviders/prometheus/prometheus.go:
```diff
@@ -140,12 +146,51 @@ func NewPrometheusProvider(api v1.API, logCtx log.Entry) *Provider {
+	for i := range keys {
+		key := keys[i]
+		formattedKey := strings.ToUpper(strings.ReplaceAll(key, "-", "_"))
+		log.Infof("formattedKey: %v", formattedKey)
```
Same as above. I feel that we should be a little conservative while using `info` level.

agrawroh commented on metricproviders/prometheus/prometheus.go:
```diff
@@ -140,12 +146,51 @@ func NewPrometheusProvider(api v1.API, logCtx log.Entry) *Provider {
+		log.Infof("formattedKey: %v", formattedKey)
+		if value, ok := os.LookupEnv(fmt.Sprintf("%s", formattedKey)); ok {
+			valuesByKey[key] = value
+			log.Infof("PROMETHEUS_ADDRESS: %v", value)
```
I'd evaluate where we can switch `info` -> `debug` to not pollute the logs.

agrawroh commented on metricproviders/prometheus/prometheus_test.go:
```diff
@@ -58,6 +62,44 @@ func TestRunSuccessfully(t *testing.T) {
+	address := "http://127.0.0.1:9090"
+	useEnvVarForKeys := true
+	os.Setenv("PROMETHEUS_ADDRESS", address)
+	tokenSecret := &corev1.Secret{
```
Do we still need this now?

agrawroh reviewed (approved):
LGTM. I'd leave the final approval to @jessesuen and @harikrongali

Thank you so much for making this change!

RaviHari commented:
> LGTM. I'd leave the final approval to @jessesuen and @harikrongali
> 
> Thank you so much for making this change!

Thanks for the review and suggestions @agrawroh @harikrongali @jessesuen

jessesuen commented on metricproviders/prometheus/prometheus.go:
```diff
@@ -140,12 +145,51 @@ func NewPrometheusProvider(api v1.API, logCtx log.Entry) *Provider {
 func NewPrometheusAPI(metric v1alpha1.Metric) (v1.API, error) {
+	secretKeys := []string{Address}
+	envValuesByKey := lookupKeysInEnv(secretKeys)
+	log.Infof("envValuesByKey: %v, len: %v", envValuesByKey, len(envValuesByKey))
```
This is a high-frequency portion of code. I prefer that we don't logging this every time we construct a prometheus API client.

jessesuen commented on metricproviders/prometheus/prometheus.go:
```diff
@@ -140,12 +145,51 @@ func NewPrometheusProvider(api v1.API, logCtx log.Entry) *Provider {
+	return err == nil && u.Scheme != "" && u.Host != ""
+}
+
+func lookupKeysInEnv(keys []string) map[string]string {
+	valuesByKey := make(map[string]string)
+	for i := range keys {
+		key := keys[i]
+		formattedKey := strings.ToUpper(strings.ReplaceAll(key, "-", "_"))
+		log.Debugf("formattedKey: %v", formattedKey)
+		if value, ok := os.LookupEnv(fmt.Sprintf("%s", formattedKey)); ok {
+			valuesByKey[key] = value
+			log.Debugf("PROMETHEUS_ADDRESS: %v", value)
+		}
+	}
+	return valuesByKey
+}
```
I don't understand the purpose of this function. Why aren't we just simply looking up an env var by its well-known, documented env var name? Can we delete this?

RaviHari replied:
This function is useful when we want to extract multiple strings from env variables such as auth headers, api keys other than just the address.. this is also converting formats to "_" in case there was a misconfiguration with "-"  in env variable .. but I agree in this case just getting from env variable is fine.

jessesuen commented on metricproviders/prometheus/prometheus.go:
```diff
@@ -22,6 +26,7 @@ const (
 	// ResolvedPrometheusQuery is used as the key for storing the resolved prometheus query in the metrics result
 	// metadata object.
 	ResolvedPrometheusQuery = "ResolvedPrometheusQuery"
+	Address                 = "PROMETHEUS_ADDRESS"
```
1. `PROMETHEUS_ADDRESS` could use a better configuration convention. I think we can go with `ARGO_ROLLOUTS_PROMETHEUS_ADDRESS`. In the future, we will have more of such environment variables for configuration and the `ARGO_ROLLOUTS` is a prefix to group rollouts specific env vars together. We do the same with `ARGOCD_`

2. nit: The global variable name `Address` is not very specific. Commonly if something is used as an env var, you can prefix it with `EnvVar<keyname>`. e.g. `EnvVarArgoRolloutsPrometheusAddress`

jessesuen commented on metricproviders/prometheus/prometheus.go:
```diff
@@ -130,7 +135,7 @@ func (p *Provider) processResponse(metric v1alpha1.Metric, response model.Value)
 }
 
-// NewPrometheusProvider Creates a new Prometheus client
+//NewPrometheusProvider Creates a new Prometheus client
```
add back space?

jessesuen reviewed (approved):
Could you please follow up this PR with doc changes?
